Python like Range (#2006)
LLNL/RAJA: docs/sphinx/user_guide/feature/iteration_spaces.rst

Thus, any iterable type that defines these methods and types appropriately
can be used as a segment with RAJA kernel execution templates.

Python-like Range Helpers
^^^^^^^^^^^^^^^^^^^^^^^^^

RAJA also provides ``RAJA::range(...)`` helpers that construct the segment
type for common half-open iteration patterns. These helpers are intended to
mirror the shape of Python's ``range`` while returning RAJA segment objects
that can be passed directly to ``RAJA::forall`` and other execution
interfaces.

The supported forms are::

  RAJA::range(end)                 // [0, end)
  RAJA::range<IndexT>(end)         // [0, end) with explicit storage type
  RAJA::range(begin, end)          // [begin, end)
  RAJA::range(begin, end, stride)  // [begin, end) with stride

The return type depends on the arguments:

* ``RAJA::range(end)`` and ``RAJA::range(begin, end)`` return a
  ``RAJA::TypedRangeSegment``.
* ``RAJA::range(begin, end, stride)`` returns a
  ``RAJA::TypedRangeStrideSegment``.
* When one of the bounds is a RAJA strong index type, such as a type created
  with ``RAJA_INDEX_VALUE``, that strong type is preserved for the loop

>   variable when possible.

Reviewer: When possible? Doesn't it always use the strong type if one is provided?
The only case I'm worried about is something like a strong index type with underlying type int combined with a long choosing long instead of being an error.

Author: that is a good point, I think we should error if so

Author: I updated the code accordingly

* Providing an explicit template argument, such as

>   ``RAJA::range<MyIndex>(end)``, overrides the deduced storage type.

Reviewer: It should still be an error to, for example, explicitly ask for int but provide a strong index type as an argument, or explicitly ask for a strong index type and provide a different strong index type as an argument.

Reviewer: I wonder if we can use c++20's require to test that these cases fail?

Reviewer 2: You definitely can, you could even make concepts to make it easier most likely.

Thus, any iterable type that defines these methods and types appropriately
can be used as a segment with RAJA kernel execution templates.

Python-like Range Helpers
^^^^^^^^^^^^^^^^^^^^^^^^^

RAJA also provides ``RAJA::range(...)`` helpers that construct the segment
type for common half-open iteration patterns. These helpers are intended to
mirror the shape of Python's ``range`` while returning RAJA segment objects
that can be passed directly to ``RAJA::forall`` and other execution
interfaces.

The supported forms are::

  RAJA::range(end)                 // [0, end)
  RAJA::range<IndexT>(end)         // [0, end) with explicit storage type
  RAJA::range(begin, end)          // [begin, end)
  RAJA::range(begin, end, stride)  // [begin, end) with stride

The return type depends on the arguments:

* ``RAJA::range(end)`` and ``RAJA::range(begin, end)`` return a
  ``RAJA::TypedRangeSegment``.
* ``RAJA::range(begin, end, stride)`` returns a
  ``RAJA::TypedRangeStrideSegment``.

> * When one of the bounds is a RAJA strong index type, such as a type created

Reviewer 3: Are this point and the next one still completely true?

Author: ah this is stale now, thanks

  with ``RAJA_INDEX_VALUE``, all bounds must use that same strong type. Mixed
  strong and plain integral bounds are rejected rather than narrowed.
* Providing an explicit template argument, such as
  ``RAJA::range<MyIndex>(end)``, overrides the deduced storage type, but
  explicit storage must still match the argument types. For example,
  ``RAJA::range<int>(RangeStrongIndex(3), 17)`` and
  ``RAJA::range<RangeStrongIndex>(AnotherRangeStrongIndex(3), 17)`` are
  rejected, while ``RAJA::range<RangeStrongIndex>(RangeStrongIndex(3),
  RangeStrongIndex(17))`` is valid.

For example::

  RAJA_INDEX_VALUE(CellIndex, "CellIndex");

  RAJA::forall<RAJA::seq_exec>(RAJA::range(N), [=](RAJA::Index_type i) {
    values[i] = i * i;
  });

  RAJA::forall<RAJA::seq_exec>(RAJA::range<CellIndex>(CellIndex {N}),
                               [=](CellIndex i) {
    typed_values[*i] = *i + 10;
  });

  RAJA::forall<RAJA::seq_exec>(RAJA::range(2, 6), [=](int i) {
    subrange_values[i] = i;
  });

  RAJA::forall<RAJA::seq_exec>(RAJA::range(CellIndex {1}, CellIndex {N},
                                            CellIndex {2}),
                               [=](CellIndex i) {

>                                  strided_values[*i] = *i;

Reviewer 3: Why dereference `i`?

Reviewer 3: If it's some feature about the CellIndex type created by the RAJA_INDEX_VALUE macro, then it might be worth mentioning in the documentation.

Author: Indeed and you are right, I think missing it that part in the RAJA documentation. Any preference to doing:  Use *idx, or RAJA::stripIndexType(idx), ? @rchen20 What is the practice in your application?

Reviewer 3: I think `*i` makes sense for this example now that you've explained it. To me it seems like `stripIndexType` is better suited to template metaprogramming since it could dereference a strong integer type or use a regular integer type directly.

Thus, any iterable type that defines these methods and types appropriately
can be used as a segment with RAJA kernel execution templates.

Python-like Range Helpers
^^^^^^^^^^^^^^^^^^^^^^^^^

RAJA also provides ``RAJA::range(...)`` helpers that construct the segment
type for common half-open iteration patterns. These helpers are intended to
mirror the shape of Python's ``range`` while returning RAJA segment objects
that can be passed directly to ``RAJA::forall`` and other execution
interfaces.

> The supported forms are::

Reviewer 3: Can you remove the extra colon?

Reviewer 4: The double colon means that the following indented lines will be formatted as code.

Author: Here it makes it a code block -- nay on the code block?

Reviewer 3: Didn't realize that made a code block.

Reviewer 4: Yup. There are multiple ways. This is the least verbose for a small snippet.

  RAJA::range(end)                 // [0, end)

>   RAJA::range<IndexT>(end)         // [0, end) with explicit storage type

Reviewer 3: The other overloads also allow explicit storage type, right? I think I would organize like so:

```
RAJA::range(end)                 // [0, end)
RAJA::range(begin, end)          // [begin, end)
RAJA::range(begin, end, stride)  // [begin, end) with stride

The storage type can be explicitly specified if desired, and the arguments will be converted to that storage type. If the argument type is not convertible to the storage type, a compilation error will occur.

RAJA::range<IndexT>(end)                 // [0, end)
RAJA::range<IndexT>(begin, end)          // [begin, end)
RAJA::range<IndexT>(begin, end, stride)  // [begin, end) with stride
```

  RAJA::range(begin, end)          // [begin, end)
  RAJA::range(begin, end, stride)  // [begin, end) with stride

The return type depends on the arguments:

* ``RAJA::range(end)`` and ``RAJA::range(begin, end)`` return a
  ``RAJA::TypedRangeSegment``.
* ``RAJA::range(begin, end, stride)`` returns a
  ``RAJA::TypedRangeStrideSegment``.
* When one of the bounds is a RAJA strong index type, such as a type created

>   with ``RAJA_INDEX_VALUE``, all bounds must use that same strong type. Mixed

Reviewer 3: I think it would be helpful to allow the following mixed case:

`RAJA::range(0, CellIndex{N}, 2)`

If there is no stride, then obviously you could shorten to `RAJA::range(CellIndex{N})`, but when you have a stride and you need a starting value, it would be nice to throw in 0 without having to wrap it since that is probably one of the most common use cases.

Author: Discussed in meeting -- we expect strongly typed loop bounds to be of the same strong type

can be used as a segment with RAJA kernel execution templates.

Python-like Range Helpers
^^^^^^^^^^^^^^^^^^^^^^^^^

RAJA also provides ``RAJA::range(...)`` helpers that construct the segment
type for common half-open iteration patterns. These helpers are intended to
mirror the shape of Python's ``range`` while returning RAJA segment objects
that can be passed directly to ``RAJA::forall`` and other execution
interfaces.

The supported forms are::

  RAJA::range(end)                 // [0, end)
  RAJA::range(begin, end)          // [begin, end)
  RAJA::range(begin, end, stride)  // [begin, end) with stride

The storage type can be explicitly specified if desired, and the arguments
are converted to that storage type. If an argument cannot be converted to the
chosen storage type, compilation fails. The explicit-type forms are::

  RAJA::range<IndexT>(end)                 // [0, end)
  RAJA::range<IndexT>(begin, end)          // [begin, end)
  RAJA::range<IndexT>(begin, end, stride)  // [begin, end) with stride

The return type depends on the arguments:

* ``RAJA::range(end)`` and ``RAJA::range(begin, end)`` return a

>   ``RAJA::TypedRangeSegment``.

Reviewer 4: ```suggestion
  ``RAJA::TypedRangeSegment<IndexT>``.
```

Better to be complete and explicit here (and below)?

examples below use ``*i`` to retrieve that wrapped value before indexing
ordinary C/C++ arrays; ``RAJA::stripIndexType(i)`` provides the same conversion
with a named helper.

For example::

  RAJA_INDEX_VALUE(CellIndex, "CellIndex");

  RAJA::forall<RAJA::seq_exec>(RAJA::range(N), [=](RAJA::Index_type i) {
    values[i] = i * i;
  });

  RAJA::forall<RAJA::seq_exec>(RAJA::range<CellIndex>(CellIndex {N}),
                               [=](CellIndex i) {
    typed_values[*i] = *i + 10;
  });

  RAJA::forall<RAJA::seq_exec>(RAJA::range(2, 6), [=](int i) {
    subrange_values[i] = i;
  });

  RAJA::forall<RAJA::seq_exec>(RAJA::range(CellIndex {1}, CellIndex {N},
                                            CellIndex {2}),
                               [=](CellIndex i) {
                                 strided_values[*i] = *i;
                               });

Strided ranges follow the same half-open interval convention as
``RAJA::TypedRangeStrideSegment``. The stride argument must have a signed
integral type. Positive strides move forward, and negative strides move
backward. For example, ``RAJA::range(N - 1, -1, -2)`` visits ``N - 1,

> N - 3, ...`` down to the first value that remains greater than ``-1``. A zero

Reviewer 4: ```suggestion
N - 3, ...`` down to the last value in the sequence that is greater than ``-1``. A zero
```

The return type depends on the arguments:

* ``RAJA::range(end)`` and ``RAJA::range(begin, end)`` return a
  ``RAJA::TypedRangeSegment<IndexT>``.
* ``RAJA::range(begin, end, stride)`` returns a

>   ``RAJA::TypedRangeStrideSegment``.

Reviewer 4: Line 131 should also have the template type specialization. I wasn't explicit about each of them in my earlier comment.

Author: thanks, I think I got them all now
